src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -191,8 +191,32 @@ private async Task PullContainerAsync(IExecutionContext executionContext, Contai
                 ArgUtil.NotNullOrEmpty(username, nameof(username));
                 ArgUtil.NotNullOrEmpty(password, nameof(password));
 
-                int loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
-                if (loginExitCode != 0)
+
+                // login with retry up to 3 times
+                int maxRetries = 3;
+                int retryCount = 0;
+                int loginExitCode = 0;
+
+                while (retryCount < maxRetries)
+                {
+                    loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
+                    if (loginExitCode == 0)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        retryCount++;
+                        if (retryCount < maxRetries)
```

Comment:
Suggestion: this check is redundant. We can leave only the "if" check in the "while" loop in line 200.

Reply:
Nice!